src/registry/otlp/attribute_otlp_ingester.rs

```diff
@@ -0,0 +1,193 @@
+
+use super::{grpc_stubs::proto::common::v1::AnyValue, listen_otlp_requests, OtlpRequest};
+
+/// An ingester for OTLP data
+pub struct AttributeOtlpIngester {
+    /// The address of the OTLP gRPC server
+    pub otlp_grpc_address: String,
+    /// The port of the OTLP gRPC server
+    pub otlp_grpc_port: u16,
+    /// The port of the admin server
+    pub admin_port: u16,
+    /// The inactivity timeout
+    pub inactivity_timeout: u64,
+}
+
+impl AttributeOtlpIngester {
+    fn maybe_to_json(value: Option<AnyValue>) -> Option<Value> {
+        if let Some(value) = value {
+            if let Some(value) = value.value {
+                use crate::registry::otlp::grpc_stubs::proto::common::v1::any_value::Value as GrpcValue;
+                match value {
+                    GrpcValue::StringValue(string) => Some(Value::String(string)),
+                    GrpcValue::IntValue(int_value) => Some(Value::Number(int_value.into())),
+                    GrpcValue::DoubleValue(double_value) => Some(json!(double_value)),
+                    GrpcValue::BoolValue(bool_value) => Some(Value::Bool(bool_value)),
+                    GrpcValue::ArrayValue(array_value) => {
+                        let mut vec = Vec::new();
+                        for value in array_value.values {
+                            if let Some(value) = Self::maybe_to_json(Some(value)) {
+                                vec.push(value);
+                            }
+                        }
+                        Some(Value::Array(vec))
+                    }
+                    _ => None,
+                }
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+}
+
+/// Iterator for OTLP attributes
+struct OtlpAttributeIterator {
+    otlp_requests: Box<dyn Iterator<Item = OtlpRequest>>,
+    buffer: Vec<SampleAttribute>,
+}
+
+impl OtlpAttributeIterator {
+    fn new(otlp_requests: Box<dyn Iterator<Item = OtlpRequest>>) -> Self {
+        Self {
+            otlp_requests,
+            buffer: Vec::new(),
+        }
+    }
+
+    fn extract_attributes_from_request(&mut self, request: OtlpRequest) -> Option<SampleAttribute> {
+        match request {
+            OtlpRequest::Logs(_logs) => {
+                // TODO Implement the checking logic for logs
+                // self.logger.error("Logs Request received");
+                self.next()
+            }
+            OtlpRequest::Metrics(_metrics) => {
+                // TODO Implement the checking logic for metrics
+                // self.logger.error("Metrics Request received");
+                self.next()
+            }
+            OtlpRequest::Traces(trace) => {
+                // Process and buffer all attributes from the trace
+                for resource_span in trace.resource_spans {
+                    if let Some(resource) = resource_span.resource {
+                        for attribute in resource.attributes {
```

Comment:
Should we also be tagging that the attribute came from a resource or span so we can use that in the rego policy?

Reply:
Yes, we'll need surrounding context. I think, when we move on to support signals properly, these requirements will pop out.